### PR TITLE
[eslint-plugin] Add attribute names rule

### PR DESCRIPTION
**Pull request by justinfagnani**
This PR should be easier to review commit-by-commit. The last commit adds the actual attribute-names rule, the rest of them set things up. I bumped typescript and a few testing dependencies due to package duplication that broke the build while installing some packages.

### Type-awareness

This implementation of attribute-names requires a type-checker. This is to detect Lit subclasses and decorators. I think this is a good tradeoff for now, especially as we can support subclasses, ReactiveElement, re-exports, etc (work on that TBD). We almost certainly do want to be able to use different heuristics to detect these things in places where they don't want to run the type-checker, so we can make some detection logic pluggable. 

typescript-eslint recommends an option or two different rules in this case:

> We recommend against changing rule logic based solely on whether `services.program` exists. In our experience, users are generally surprised when rules behave differently with or without type information. Additionally, if they misconfigure their ESLint config, they may not realize why the rule started behaving differently. Consider either gating type checking behind an explicit option for the rule or creating two versions of the rule instead.

https://typescript-eslint.io/developers/custom-rules/#typed-rules

We'll have to figure that out.

### Copyright assignment

@43081j I originally copied the rule from https://github.com/43081j/eslint-plugin-lit/blob/master/src/rules/attribute-names.ts but ended up rewriting most of it to work against the Lit analyzer. I did copy the tests, then added the correct Lit imports to satisfy the analyzer's detection of LitElement and decorators.

The commits to this repo need to be copyright Google (unless they're a compatible OSS license, credited, and linked). I think the rule implementation qualifies. The test is written by you and you've signed the CLA, so I believe you can just reassign copyright to Google at will with your approval here. Or I can write new tests if necessary.

### CODE REVIEW

File: packages/labs/eslint-plugin/src/lib/util.ts
Comment (augustjk) [inline packages/labs/eslint-plugin/src/lib/util.ts:18]: is this worth a TODO or an issue for actually adding docs to lit.dev ?

File: packages/labs/eslint-plugin/src/lib/util.ts
Comment (augustjk) [inline packages/labs/eslint-plugin/src/lib/util.ts:32]: should this be removed?

File: packages/labs/eslint-plugin/src/lib/util.ts
Comment (augustjk) [inline packages/labs/eslint-plugin/src/lib/util.ts:59]:
```suggestion
  e.kind === ts.SyntaxKind.FalseKeyword;
```

File: packages/labs/eslint-plugin/src/rules/attribute-names.ts
Comment (augustjk) [inline packages/labs/eslint-plugin/src/rules/attribute-names.ts:58]: should this be removed?

File: packages/labs/testing/package.json
Comment (augustjk) [inline packages/labs/testing/package.json:111]: nothing that this change warrants a bump to the labs/testing package.

File: packages/labs/eslint-plugin/package.json
Comment (augustjk) [inline packages/labs/eslint-plugin/package.json:39]: is narrowing the --test glob intentional?

**Review by augustjk — COMMENTED** [review]
Commenting minor things that stood out to me. I'll defer to other reviewers for the actual rule implementation.
Additionally noting that packages like labs/cli and labs/compiler that have direct dependency on typescript will also need version bumps with this.

**Comment by augustjk** [comment]
It is curious that this has size regression for core packages. Could changing the TS version change the emitted JS?

File: packages/labs/eslint-plugin/src/lib/util.ts
Comment (justinfagnani) [reply packages/labs/eslint-plugin/src/lib/util.ts:18]: ah, yes

**Comment by justinfagnani** [comment]
> It is curious that this has size regression for core packages. Could changing the TS version change the emitted JS?

@augustjk I was wondering the same thing 🤷‍♂️

File: packages/labs/eslint-plugin/package.json
Comment (justinfagnani) [reply packages/labs/eslint-plugin/package.json]: nope!

**Comment by 43081j** [comment]
> typescript-eslint recommends an option or two different rules in this case

i would have two rules i think, a typed and untyped one (which share some logic). then just publish a typed/untyped recommended config

> The test is written by you and you've signed the CLA, so I believe you can just reassign copyright to Google at will with your approval here

sounds good to me.

i'm not too concerned with the copyright, copying things over here, etc. as long as im still able to exist as a maintainer alongside you all (which it seems i am)

File: package-lock.json
Comment (augustjk) [inline package-lock.json:23670]: Looks like this terser update might be responsible for the bundle size change. I tried updating typescript alone and it didn't change the output but this branch does have a diff. Diff looks innocuous. Somehow previous version lucked out and didn't require an alias for `H`, but after update it does.
image

**Review by 43081j — APPROVED** [review]
looks good to me 🥳

File: packages/labs/eslint-plugin/src/test/rules/attribute-names_test.ts
Comment (rictic) [inline packages/labs/eslint-plugin/src/test/rules/attribute-names_test.ts:102]: Huh, I hadn't seen this rule before. Do we like this rule? Do we want to recommend it?

The casedAttribute message seems good, but the casedPropertyWithoutAttribute doesn't seem great to me.

File: packages/tests/src/web-test-runner.config.ts
Comment (rictic) [inline packages/tests/src/web-test-runner.config.ts:150]: where did these new flags come from?

File: packages/labs/eslint-plugin/src/index.ts
Comment (rictic) [inline packages/labs/eslint-plugin/src/index.ts:9]: Should this be empty?

Comment (justinfagnani) [reply packages/labs/eslint-plugin/src/index.ts:9]: nope!

File: packages/tests/src/web-test-runner.config.ts
Comment (justinfagnani) [reply packages/tests/src/web-test-runner.config.ts:150]: I found these in a few places that were documenting trying to work around Chrome's background tab throttling. I'm trying to serperate and/or revert the changes causing the problems to unblock this PR.

File: packages/labs/eslint-plugin/src/test/rules/attribute-names_test.ts
Comment (justinfagnani) [reply packages/labs/eslint-plugin/src/test/rules/attribute-names_test.ts:102]: https://github.com/43081j/eslint-plugin-lit/blob/master/docs/rules/attribute-names.md

I'm planning on porting all the existing rules to ease migration to the new version. I'm sure there is some overlap with Rune's rules, and some we might want to deprecate going forward. I agree that a lot of perfectly fine code may have a camelCased property name and be making the choice to use the lower-case / case insensitive attribute name and shouldn't be flagged. In those cases we may want additional configuration for this rule, and we may want a migration-oriented configuration and not include this in the recommended set.

I should open an RFC on how the plan for all of this.

**Pull request merged**
